### ISSUE TITLE
The mpi problem

### ISSUE THREAD

**Issue by lev87** [Microsoft/CNTK#3056]
My GPU have 4 cards,  when I use the command: mpiexec -np 4 cntk  xxxx deviceId=auto. There are only 1 or 2 processes, and some processes are stucked. when I set for example deviceId=0 ,there will be 4 processes on one card. I want to run 4 processes on all the cards, how can I do?

**Comment by jaliyae** [comment]
Could you try 'deviceid=all' option and repeat the experiement.

**Comment by lev87** [comment]
@jaliyae  Thanks reply, I use cntk 2.3, when try deviceid=all option, there will be an EXCEPTION: invalid  input string 'all'

**Comment by jaliyae** [comment]
Could you try running a python example in your machine to see if this is specific to the brainscript version?

**Comment by lev87** [comment]
@jaliyae Yes, python example works well

**Comment by jaliyae** [comment]
I don't think this value is transferred correctly in the brainscript evaluation. The evaluation probably use the default device and other MPI processes fail because it cannot lock the device it needs. One option is to write a small wrapper to set the device id in MPI processes.

**Comment by KeDengMS** [comment]
If you are running on Linux, please check if you have pending lock files like /var/lock/CNTK_exclusive_lock_for_GPU_*.
Windows uses a global mutex for GPU locks, and reboot should clean them up.

**Comment by lev87** [comment]
I don't have the root authority of the computer, so I don't set the gpu lock file with /var/lock when compile. Will this have a difference?

**Comment by KeDengMS** [comment]
BrainScript training should automatically select different GPUs without deviceId=auto. If distributed training in Python works, BrainScript should work as well. If you are running in Linux, please try cntk docker images from docker hub.
